tests/rest/client/test_auth.py

```diff
@@ -624,6 +624,51 @@ def test_refreshable_access_token_expiration(self):
             refresh_response.json_body["expires_in_ms"], 60 * 1000, 100
         )
 
+    @override_config({"access_token_lifetime": "1m", "refresh_token_lifetime": "2m"})
+    def test_refresh_token_expiry(self):
+        """
+        The refresh token can be configured to have a limited lifetime.
+        """
+
+        def use_refresh_token(refresh_token: str) -> FakeChannel:
+            """
+            Helper that makes a request to use a refresh token.
+            """
+            return self.make_request(
+                "POST",
+                "/_matrix/client/unstable/org.matrix.msc2918.refresh_token/refresh",
+                {"refresh_token": refresh_token},
+            )
+
+        body = {"type": "m.login.password", "user": "test", "password": self.user_pass}
+        login_response = self.make_request(
+            "POST",
+            "/_matrix/client/r0/login?org.matrix.msc2918.refresh_token=true",
+            body,
+        )
+        self.assertEqual(login_response.code, 200, login_response.result)  #
+        refresh_token = login_response.json_body["refresh_token"]
+
+        # Advance 119 seconds in the future (just shy of 2 minutes)
+        self.reactor.advance(119.0)
+
+        # Refresh our access token. It should still JUST be valid right now.
```

Comment:
Also, if I understand correctly:
- at this point the access token has expired, so you shouldn't be able to use it
- using the refresh token will allow us to use the access token again(?)
    - If so, do we test this?

Reply:
You're right in saying there should be more tests.

> using the refresh token will allow us to use the access token again(?) 

It creates a *new* access token :). It's actually pretty tricky to invalidate the caches on other Synapse workers to extend a token's lifetime (I looked into this before we switched track to refresh tokens for a customer), it's actually easier to create a fresh token. (Also, creating a new token is kind of how refresh tokens protect against access token disclosure, since the leaked access token dies out and never is valid again.)

Reply:
> Shouldn't it be valid for a fresh batch of 2 minutes? Otherwise, why are you waiting 121 seconds below rather than 2 seconds?

Yes. I think there must be something unclear about the test since I think the test is right, though I'm not sure I understand what you're asking.

Here, we first test that a refresh (of R1) works within 119 seconds of issuance (less than 2 minutes; this produces R2) but a refresh (of R2) fails after 121 seconds of issuance (of R2).

Perhaps one thing that would help would be to not reuse the same variable name.

I will tweak some comments / names and see if it pleases you more

Reply:
Thanks, I think those changes are good. What I hadn't groked was that on line +659 we now have a second (new) access token.

The phrase "refresh our access token" on +671 makes me think of keeping the same access token around, but delaying its expiry. Perhaps it would be clearer to speak of "requesting a new access token" or "refreshing/extending the session" here and in #11427?

Reply:
Thanks, that is clearer!